### ISSUE TITLE
SwissProt and TrEMBL vs. UniProt Knowledgebase

### ISSUE THREAD

**Issue by ariutta** [bridgedb/BridgeDb#17]
Our [datasources.txt](https://github.com/bridgedb/BridgeDb/blob/master/org.bridgedb.bio/resources/org/bridgedb/bio/datasources.txt) official names (last column) don't always match the recommended Miriam names. For example, datasources.txt has "UniProt/Trembl" where Miriam has ["UniProt Knowledgebase"](http://identifiers.org/uniprot/). (Update: not an exact match. See later comments.)

**Comment by egonw** [comment]
@ariutta, @Chris-Evelo, @AlexanderPico the MIRIAM name does not reflect the subset... same applies for the Swissprot subset. Is that OK? Or should the proper names be "UniProt Knowledgebase/TrEMBL" (etc)?

**Comment by Christian-B** [comment]
The way I have always understood it. (but may be wrong) is.
Things start put as Trembl when they are automatically added and given a tremble URL,
When they are then later manually verified they are added to Swissport and given a DIFFERENT Swissprot url (Where different is both the datasource and ID part)
However most people including OPS continue to use the Trembl URLs but only included in their data the objects which have been added to the Swissprot dataset.
So in BredgeBb/ IMS terms the correct term is Trembl and NOT swissprot as it is still the Trembl URLs that are being used.  In fact BridgeDb would have no way of knowing automatically looking just at the ID which is which.

Indentifiers.org/Miriam also ONLY do regex pattern matching to see if a URL is valid. There is no regex rules to split TREMBL only URLs from Trembl URsfor items that are included in the Swissprot subset. So they will never be able to make the distinction.

**Comment by ariutta** [comment]
@egonw, you're right -- "UniProt Knowledgebase" refers to both Swissprot and TrEMBL. When I opened this issue I incorrectly thought "http://identifiers.org/uniprot" only referred to TrEMBL, but here's part of an email reply from Nick Juty:

> ... I've updated the record for TrEMBL synonyms as requested. Our record does actually refer to SwissProt as well. I guess you want those added too? We go by what is listed on the UniProt website in this case: http://www.ebi.ac.uk/uniprot. It states 1) 'The  UniProt Knowledgebase (UniProtKB) ...consists of two sections:
    UniProtKB/Swiss-Prot  which is manually annotated and is reviewed and
    UniProtKB/TrEMBL  which is automatically annotated and is not reviewed.'

Since "http://identifiers.org/uniprot" refers to both TrEMBL and Swissprot, I don't know how we should indicate the official name and the Miriam URN in datasources.txt. I'm open to opinions on whether "UniProt Knowledgebase/TrEMBL" is best and whether "urn:miriam:uniprot" should be listed for 1) both TrEMBL and Swissprot or 2) just TrEMBL.

**Comment by Chris-Evelo** [comment]
For me the important question is whether we can up with a method to use URN's that are both technically correct and that allow biologists to actually judge immediately from that URN whether data is evaluated (as it is when from UniProt) or not. We might have to go back to identifiers.org or the UniProt team for that.

**Comment by Christian-B** [comment]
Hi Chris,

To the best of my knowledge the is no way by looking at just the URL/URN even the uniprot one  if the object has been added to swissprot or is just Trembl, especially as this could change over time.  The biologist would have to look the data up at uniprot.  

identifiers.org do not keep individual data records. They only store ID regex patterns passing all id level calls down to the underlying data source (in this case uniport)

**Comment by Chris-Evelo** [comment]
Just thinking out loud now. But things that are in SwissProt actually have a SwissProt ID too right? And the UniProt entry should (and probably will) link to that. Can we use that somehow?

**Comment by Christian-B** [comment]
Yes as far I know things in Swissprot have a second completely different SwissProtID having a linkset to that would be nice but never done in OPS.
Note the SwissProt URN is different in both the base and ID part so no semi automated system could be used,

**Comment by ariutta** [comment]
Another comment from Nick Juty:

>I had a very quick look. I would map both swiss-prot and trembl to our Uniprot collection, as that is what it reflects. I'm not sure of how this causes issues, aside from not knowing from the URI itself which is a manually curated record, and which is not. I've also added UniProtKB/Swiss-Prot as a synonym in case that helps you a bit?

But if datasources.txt is correct that `CALM_HUMAN` is a SwissProt identifier, then the http://identifiers.org/uniprot identifier pattern:

```
^([A-N,R-Z][0-9]([A-Z][A-Z, 0-9][A-Z, 0-9][0-9]){1,2})|([O,P,Q][0-9][A-Z, 0-9][A-Z, 0-9][A-Z, 0-9][0-9])(\.\d+)?$
```

is incorrect, because it [does not match](http://regexr.com/3ce3p) `CALM_HUMAN`.

**Comment by ariutta** [comment]
Currently identifiers.org has "UniProt Knowledgebase" as the preferred name and additionally has these alternative names listed for the [http://identifiers.org/uniprot/](http://identifiers.org/uniprot/) data collection:
* UniProtKB
* UniProt
* Protein Knowledgebase
* UniProt-TrEMBL
* UniProt/TrEMBL
* UniProtKB/Swiss-Prot

If you want me to suggest any additions or changes, just let me know.

**Comment by ariutta** [comment]
Followup comment from Nick:

>I think that the gene name is CALM, and in humans it is CALM_HUMAN. But we use the identifier provided by UniProt for the record, not for the gene or protein. The identifier for the record is the stable identifier.

So is datasources.txt wrong to list `CALM_HUMAN` as a sample identifier for SwissProt?

**Comment by AlexanderPico** [comment]
Just to throw another wrench into this discussion… The datasource names and identifiers used by BridgeDb are not only reliant on our collective best sense of what it *should* be, nor on what identifiers.org <http://identifiers.org/> does, but also critically dependent on what the primary resources such as Ensembl decide to do.  The BridgeDb database build has been simplified over many years to depend on resources like Ensembl to make these decisions, since they dedicate a ton of time to the problem and represent a widely-use community resource.  In many ways this relieves us of having to figure this out and make the necessarily compromising decisions. 

In other words, let’s just do what Ensembl does and it simplifies BOTH the decision making and the build process.

If something about how the source data from Ensembl is really really offensive, or caused specific data integrity issues, or breaks a critical analysis workflow, well then we should take it up with Ensembl and ask if they can change it.  The more we drift from them, the more we have to maintain these differences in code and in practice.

My perspective on keeping things simple :)
 - Alex

**Comment by ariutta** [comment]
@AlexanderPico, keeping things simple sounds good. I took a look at the Ensembl entry for [CALM2 ](http://identifiers.org/ensembl/ENSG00000143933), and here's how they refer to Uniprot:

>UniProtKB This gene has proteins that correspond to the following Uniprot identifiers: [P62158](http://www.uniprot.org/uniprot/P62158)

They don't appear to distinguish between SwissProt and TrEMBL.

**Comment by AlexanderPico** [comment]
Ah, not the Ensembl website, but rather their databases. Specifically the data tables we use to make the bridgedb database. This is different from their singular representation. It covers all the alias representations, including uniprot. 

 - Alex

**Comment by egonw** [comment]
Closed by https://github.com/bridgedb/BridgeDb/commit/192a18f2ee626279b0845a597b5861f02fe8e521
